org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/handler/ZigBeeThingHandler.java

```diff
@@ -175,18 +180,43 @@ private synchronized void doNodeInitialisation() {
 
         logger.debug("{}: Start initialising ZigBee Thing handler", nodeIeeeAddress);
 
+        // Update the general properties
+        // This is required here to allow us to check for a static thing type before we check the dynamic configuration
+        ZigBeeNodePropertyDiscoverer propertyDiscoverer = new ZigBeeNodePropertyDiscoverer();
+        propertyDiscoverer.setProperties(getThing().getProperties());
+        Map<String, String> newProperties = propertyDiscoverer.getProperties(node);
+        updateProperties(newProperties);
+
         // Create the channel factory
         ZigBeeChannelConverterFactory factory = new ZigBeeChannelConverterFactory();
+        List<Channel> nodeChannels;
+
+        // Check the discovery data to see if we match this thing with a static definition
+        ZigBeeThingTypeMatcher matcher = new ZigBeeThingTypeMatcher();
+        ThingTypeUID thingTypeUid = matcher.matchThingType(newProperties);
```

Comment:
I was wondering why the thing type matching using discovery participants is not done here as well, but only in the discovery service. What is the reason for this? Wouldn't it be useful to identify the static thing type here as well using all available discovery participants?

Reply:
How can this be done - the list of discovery participants are only known in the discovery handler, and not the thing handler? In fact, this code probably should be removed now as the thing type discovery is performed in the discovery handler.

Reply:
My comment was inspired by my memory of a discussion that we had some months ago, when this PR was initially opened. Back then, I was wondering why the thing type matching is done in the thing handler, and not in the discovery. I recall your (for me then convincing :) ) answer to be that a discovery result might be approved before the node properties have been read from the ZigBee device, and that, therefore, doing the thing type matching in the discovery might be too early.

Maybe I overlook something, but can't this be still the case now with thing type matching being done by the 'participants'?

Reply:
> Back then, I was wondering why the thing type matching is done in the thing handler, and not in the discovery.

Yes, I recall this, but now everything has changed and the discovery can now only be done in the discovery handler as it is the only place that knows the discovery participants which are doing the matching of thing types to the node.

> can't this be still the case now with thing type matching being done by the 'participants'?

Maybe I'm missing something, but I don't see how? The ThingHandler does not know anything about the discovery participants, so how can it discover thing types? To do what you suggest, we would need to inject the participants list into the thing handler right?

Reply:
Yes, one would need to inject the participants into the thing handler (and - probably - modify the discovery participant interface to make them usable from within the thing handler, maybe resulting in something that looks similar to an interface extracted from the `ThingTypeMatcher` class).

Anyhow, I am currently not sure whether this is worth changing - the only "disadvantage" that I see from having the extensible thing type matching only in the discovery (and not in the thing handler) is the following:
* If a discovery result is approved before the node properties are available, then the resulting thing would stay a 'generic' ZigBee device, and would not get a non-generic thing type.

I could live with that.

Reply:
> one would need to inject the participants into the thing handler

Is that allowed? If we do this, we have the thing handler relying on the discovery service which I thought we were trying to keep separated?

> If a discovery result is approved before the node properties are available, then the resulting thing would stay a 'generic' ZigBee device, and would not get a non-generic thing type.

Yes - this is why I originally designed the system in the way it was designed (as I think we discussed a while back).

I would probably suggest to leave this out for now as it seems quite messy to me if you don't feel too strongly about it...

Reply:
I do not feel too strongly about it (the problem is not relevant to my usage of the binding - I will only approve discovery results after a non-generic thing type was found).

I have ideas how this issue could be resolved; if you want, we can discuss, but we can also leave this out, I am fine with that.